File: backend/src/main/java/com/protractor/backend/domain/posture/dto/PostureFeatures.java
package com.protractor.backend.domain.posture.dto;

import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.validation.Valid;
import jakarta.validation.constraints.DecimalMax;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.NotNull;

/**
 * 자세 피처 벡터 v1. 브라우저가 MediaPipe 키포인트에서 뽑아 보내는 값이며, 판정 방식 3종(규칙·하이브리드·이미지)이 모두 이
 * 벡터를 공통 입력으로 쓴다.
 *
 * <p>
 * 원본 좌표를 그대로 쓰면 카메라 거리·해상도·체형에 따라 값이 제각각이라 임계값을 정할 수 없다. 그래서 브라우저에서 <b>어깨 중점을
 * 원점으로 옮기고 어깨 너비로 나눈</b> 무차원 비율만 보낸다. 덕분에 전송량도 작아서 1Hz로 보내도 부담이 없다.
 *
 * <p>
 * 이 필드 구성은 v1으로 동결한다. 방법 A/B/C의 정확도를 비교하려면 세 방식이 같은 입력을 받아야 하므로, 필드를 바꿀 때는 v2를 새로
 * 만들고 데이터셋도 함께 다시 수집해야 한다.
 *
 * <p>
 * {@code mlFeatures}는 그 동결을 깨지 않으려고 <b>덧붙인</b> 선택 블록이다. 학습 모델이 v1에 없는 값(눈 좌표·깊이)을
 * 쓰기 때문에 필요한데, v1 필드를 건드리면 규칙 기반 판정기와 기존 캘리브레이션 기준선이 모두 무효가 된다. 이 블록이 비어 있어도 판정은
 * 예전과 똑같이 동작한다.
 *
 * @param capturedAtMillis 브라우저 기준 촬영 시각(epoch ms). 서버 수신 시각과 다를 수 있어 함께 받는다
 * @param earShoulderOffsetRatio 귀와 어깨 중점의 <b>수평</b> 거리 / 어깨 너비. 거북목 주지표(클수록 앞으로
 *     나옴)
 * @param earShoulderVerticalRatio 귀와 어깨의 <b>수직</b> 거리 / 어깨 너비. 거북목 각도 계산의 분모
 * @param shoulderToFaceWidthRatio 어깨 너비 / 얼굴 너비. 라운드숄더 주지표(어깨가 말리면 정면 폭이 줄어 값이
 *     작아진다). 얼굴 너비로 나누는 이유는 카메라와의 거리가 변해도 비율이 유지되기 때문이다
 * @param shoulderTiltRatio (왼쪽 어깨 y - 오른쪽 어깨 y) / 어깨 너비. 어깨 높낮이 주지표. <b>부호를
 *     유지</b>해서 어느 쪽이 올라갔는지 구분한다
 * @param torsoRotationRatio 좌우 어깨가 코에서 떨어진 거리의 차이 / 어깨 너비. 몸통이 얼마나 틀어졌는지 추정한다.
 *     회전에 취약한 자세는 이 값이 크면 판정을 보류한다
 * @param neckFlexionRatio 코와 어깨 중점의 수직 거리 / 어깨 너비. 고개 숙임 정도(보조 지표)
 * @param faceWidthRatio 얼굴 너비 / 어깨 너비. 정규화가 정상인지 확인하는 검증용 값
 * @param earVisibility 좌우 귀 랜드마크 신뢰도의 최솟값(0~1). 낮으면 거북목 판정을 보류한다
 * @param shoulderVisibility 좌우 어깨 랜드마크 신뢰도의 최솟값(0~1). 낮으면 어깨 관련 판정을 보류한다
 * @param mlFeatures 학습 모델 전용 피처 블록(선택). 하이브리드 판정기만 쓰며, 없으면 규칙 기반으로 판정한다
 */
@Schema(description = "자세 피처 벡터 v1(브라우저에서 정규화 완료)")
public record PostureFeatures(
		@Schema(description = "촬영 시각(epoch ms)", example = "1769472000000") @NotNull Long capturedAtMillis,
		@Schema(description = "귀-어깨 수평거리 / 어깨너비", example = "0.18") @NotNull Double earShoulderOffsetRatio,
		@Schema(description = "귀-어깨 수직거리 / 어깨너비", example = "0.62") @NotNull Double earShoulderVerticalRatio,
		@Schema(description = "어깨너비 / 얼굴너비", example = "2.35") @NotNull Double shoulderToFaceWidthRatio,
		@Schema(description = "좌우 어깨 높이차 / 어깨너비(부호 유지)", example = "-0.04") @NotNull Double shoulderTiltRatio,
		@Schema(description = "몸통 회전 추정치", example = "0.09") @NotNull Double torsoRotationRatio,
		@Schema(description = "코-어깨중점 수직거리 / 어깨너비", example = "0.71") @NotNull Double neckFlexionRatio,
		@Schema(description = "얼굴너비 / 어깨너비", example = "0.43") @NotNull Double faceWidthRatio,
		@Schema(description = "귀 랜드마크 신뢰도(0~1)", example = "0.93") @NotNull @DecimalMin("0.0") @DecimalMax("1.0") Double earVisibility,
		@Schema(description = "어깨 랜드마크 신뢰도(0~1)", example = "0.97") @NotNull @DecimalMin("0.0") @DecimalMax("1.0") Double shoulderVisibility,
		@Schema(description = "학습 모델 전용 피처(선택). 하이브리드 판정기에서만 쓴다") @Valid PostureMlFeatures mlFeatures) {

	/** ML 블록 없이 v1 필드만으로 만든다. 규칙 기반 판정과 테스트에서 쓴다. */
	public PostureFeatures(Long capturedAtMillis, Double earShoulderOffsetRatio, Double earShoulderVerticalRatio,
			Double shoulderToFaceWidthRatio, Double shoulderTiltRatio, Double torsoRotationRatio,
			Double neckFlexionRatio, Double faceWidthRatio, Double earVisibility, Double shoulderVisibility) {
		this(capturedAtMillis, earShoulderOffsetRatio, earShoulderVerticalRatio, shoulderToFaceWidthRatio,
				shoulderTiltRatio, torsoRotationRatio, neckFlexionRatio, faceWidthRatio, earVisibility,
				shoulderVisibility, null);
	}

	/** 피처 벡터 버전. events.metadata에 함께 남겨 어떤 규격으로 판정했는지 나중에 알 수 있게 한다. */
	public static final String VERSION = "v1";
}

File: frontend/src/lib/pose/postureFeatures.ts
// 서버 자세 판정에 보내는 피처 추출.
//
// ⚠️ 이 파일은 backend 의 PostureFeatures / PostureMlFeatures 계약과 ai/train.ipynb 의
// compute_features 를 그대로 옮긴 것이다. 한 줄이라도 어긋나면 서버는 예외 없이
// "그럴듯하게 틀린" 판정을 내놓고, 그건 테스트 없이는 발견되지 않는다.
//
// landmarks.ts 의 computeMetrics 와 계산이 겹쳐 보이지만 일부러 합치지 않았다.
// computeMetrics 는 입장 준비 화면 전용이라 정의가 다르다.
//   - 눈 랜드마크가 EYE_OUTER(3·6)인데 학습 모델은 EYE(2·5)를 쓴다
//   - 어깨 기울기를 각도(도)로 내는데 서버 계약은 어깨너비로 나눈 비율이다
//   - 깊이(z)를 아예 보지 않는다
// 둘을 합치면 학습 때와 다른 값이 서버로 간다.
import type { NormalizedLandmark } from '@mediapipe/tasks-vision';

// 
import type {
  PostureBaseline,
  PostureFeatures,
  PostureMlFeatures,
} from '@/types/posture';

/** BlazePose 랜드마크 인덱스 — 서버 계약이 쓰는 것만 */
const NOSE = 0;
const L_EYE = 2;
const R_EYE = 5;
const L_EAR = 7;
const R_EAR = 8;
const L_SHOULDER = 11;
const R_SHOULDER = 12;

interface P2 {
  x: number;
  y: number;
  v: number;
}

interface P3 {
  x: number;
  y: number;
  z: number;
}

const dist2 = (a: P2 | P3, b: P2 | P3) => Math.hypot(a.x - b.x, a.y - b.y);

/**
 * 피처 벡터 v1 을 만든다. 어깨나 얼굴이 안 잡히면 null.
 *
 * MediaPipe 의 normalized 좌표는 x 가 이미지 너비, y 가 높이 기준이라 종횡비가 1이 아니면
 * 단위가 서로 다르다. 그래서 픽셀 좌표로 되돌려 계산한다.
 */
export function extractPostureFeatures(
  lm: NormalizedLandmark[],
  width: number,
  height: number,
): PostureFeatures | null {
  const P = (i: number): P2 => ({
    x: lm[i].x * width,
    y: lm[i].y * height,
    v: lm[i].visibility ?? 0,
  });
  const mid = (a: P2, b: P2) => ({ x: (a.x + b.x) / 2, y: (a.y + b.y) / 2, v: 0 });

  const nose = P(NOSE);
  const lEar = P(L_EAR);
  const rEar = P(R_EAR);
  const lSh = P(L_SHOULDER);
  const rSh = P(R_SHOULDER);

  const shoulderWidth = dist2(lSh, rSh);
  const faceWidth = dist2(lEar, rEar);
  if (shoulderWidth <= 0 || faceWidth <= 0) return null;

  const shMid = mid(lSh, rSh);
  const earMid = mid(lEar, rEar);

  return {
    capturedAtMillis: Date.now(),
    // 귀-어깨 중점의 수평/수직 거리. 서버는 거북목 각도를 atan2(수평, 수직)로 낸다
    earShoulderOffsetRatio: Math.abs(earMid.x - shMid.x) / shoulderWidth,
    earShoulderVerticalRatio: Math.abs(earMid.y - shMid.y) / shoulderWidth,
    // 어깨가 앞으로 말리면 정면 어깨 폭이 줄어든다. 카메라 거리 변화를 지우려고 얼굴 너비로 나눈다
    shoulderToFaceWidthRatio: shoulderWidth / faceWidth,
    // 어느 쪽 어깨가 올라갔는지 알아야 해서 부호를 유지한다
    shoulderTiltRatio: (lSh.y - rSh.y) / shoulderWidth,
    // 몸통 회전: 코에서 좌우 어깨까지 거리 차이. 크면 서버가 어깨 판정을 보류한다
    torsoRotationRatio:
      Math.abs(dist2(nose, lSh) - dist2(nose, rSh)) / shoulderWidth,
    neckFlexionRatio: Math.abs(nose.y - shMid.y) / shoulderWidth,
    faceWidthRatio: faceWidth / shoulderWidth,
    // 서버는 최솟값 기준으로 보류를 판단한다. 여기서 미리 최솟값을 낸다
    earVisibility: Math.min(lEar.v, rEar.v),
    shoulderVisibility: Math.min(lSh.v, rSh.v),
    mlFeatures: extractMlFeatures(lm, width, height),
  };
}

/**
 * 학습 모델 전용 피처 12개 — ai/train.ipynb 의 compute_features 와 100% 같아야 한다.
 *
 * v1 피처와 따로 계산하는 이유는 두 가지다.
 * 1) 학습 모델이 v1 에 없는 값을 쓴다 — 눈 좌표(눈 간격·머리 기울기)와 MediaPipe 의 z(깊이).
 * 2) 겹치는 4개도 정의가 미묘하게 다르다. v1 은 절댓값을 쓰거나(neckFlexionRatio)
 *    부호가 반대라서(shoulderTiltRatio), v1 에서 유도하면 학습 때와 다른 값이 들어간다.
 *
 * z 는 학습 때와 같은 스케일이어야 해서 x 와 같은 기준(이미지 너비)으로 되돌린다.
 *
 * 반환 순서는 노트북 FEATURE_ORDER 와 맞춰 둔다. 서버는 이름으로 꺼내 쓰므로 순서가 결과를
 * 바꾸지는 않지만, 목록을 나란히 놓고 대조할 수 있어야 피처를 빠뜨리지 않는다.
 */
export function extractMlFeatures(
  lm: NormalizedLandmark[],
  width: number,
  height: number,
): PostureMlFeatures | null {
  const P = (i: number): P3 => ({
    x: lm[i].x * width,
    y: lm[i].y * height,
    z: (lm[i].z ?? 0) * width,
  });
  const mid = (a: P3, b: P3): P3 => ({
    x: (a.x + b.x) / 2,
    y: (a.y + b.y) / 2,
    z: (a.z + b.z) / 2,
  });

  const nose = P(NOSE);
  const lEye = P(L_EYE);
  const rEye = P(R_EYE);
  const lEar = P(L_EAR);
  const rEar = P(R_EAR);
  const lSh = P(L_SHOULDER);
  const rSh = P(R_SHOULDER);

  const shoulderWidth = dist2(lSh, rSh);
  const eyeDist = dist2(lEye, rEye);
  // 서버는 블록이 통째로 있거나 없거나만 받는다. 하나라도 못 만들면 null 을 보내 규칙 기반으로 넘긴다
  if (shoulderWidth < 1e-6 || eyeDist < 1e-6) return null;

  const shMid = mid(lSh, rSh);
  const earMid = mid(lEar, rEar);
  const earShoulderDy = earMid.y - shMid.y;
  const earShoulderDz = earMid.z - shMid.z;

  const headRoll = (rEye.y - lEye.y) / eyeDist;
  const shoulderTilt = (rSh.y - lSh.y) / shoulderWidth;

  return {
    eyeDistRatio: eyeDist / shoulderWidth,
    earDistRatio: dist2(lEar, rEar) / shoulderWidth,
    noseAboveShoulder: (shMid.y - nose.y) / shoulderWidth,
    earAboveShoulder: (shMid.y - earMid.y) / shoulderWidth,
    noseZRel: (nose.z - shMid.z) / shoulderWidth,
    earZRel: earShoulderDz / shoulderWidth,
    // 귀-어깨 측면(수직+깊이) 2D 거리 근사. 선형결합으로는 못 만드는 신호라 따로 넣는다
    acromionProxy: Math.hypot(earShoulderDy, earShoulderDz) / shoulderWidth,
    headRoll,
    shoulderTilt,
    // 어느 쪽으로 기울든 나쁜 자세다. 선형 모델은 부호 있는 값 하나로 그걸 표현할 수 없다
    absShoulderTilt: Math.abs(shoulderTilt),
    absHeadRoll: Math.abs(headRoll),
    // 좌우 어깨의 깊이 차. 몸통을 튼 경우에 커지므로 라운드숄더와 회전을 구분한다
    shoulderZSpread: Math.abs(lSh.z - rSh.z) / shoulderWidth,
  };
}

/**
 * 캘리브레이션 기준선 = 수집한 피처의 평균. 서버 PostureBaseline 이 쓰는 v1 4개와,
 * 델타 판정기용 ML 피처 12개의 평균을 함께 낸다.
 *
 * 평균이 아니라 중앙값을 쓰지 않는 이유는 서버 계약(sampleCount 를 함께 받는 평균값)과
 * 맞추기 위해서다. 대신 표본이 적으면 기준선을 믿을 수 없으므로 호출부에서 최소 개수를 확인한다.
 */
export function averageBaseline(samples: PostureFeatures[]): PostureBaseline {
  const avg = (pick: (f: PostureFeatures) => number) =>
    samples.reduce((sum, f) => sum + pick(f), 0) / samples.length;

  return {
    version: 'v1',
    earShoulderOffsetRatio: avg((f) => f.earShoulderOffsetRatio),
    earShoulderVerticalRatio: avg((f) => f.earShoulderVerticalRatio),
    shoulderToFaceWidthRatio: avg((f) => f.shoulderToFaceWidthRatio),
    shoulderTiltRatio: avg((f) => f.shoulderTiltRatio),
    sampleCount: samples.length,
    mlBaseline: averageMlBaseline(samples),
  };
}

/**
 * ML 피처 12개의 평균. hybrid-delta 판정기가 델타를 만드는 피감수다.
 *
 * mlFeatures 가 없는 프레임(랜드마크 부족)은 빼고 평균을 낸다. 0 으로 채워 넣으면 평균이
 * 통째로 어긋나고, 그건 서버에서 "이 사람의 바른 자세"로 쓰이므로 판정이 조용히 틀어진다.
 * 하나도 없으면 null 을 보내고 서버가 델타 판정을 보류하게 둔다 — 없는 기준선을 지어내는 것보다 낫다.
 *
 * ⚠️ 학습 노트북(ai/train.ipynb 의 build_person_baselines)도 같은 정의를 쓴다.
 * 그쪽은 "바른 자세 영상 프레임의 피처 평균"이고 여기는 "캘리브레이션 구간의 피처 평균"이다.
 * 한쪽만 바꾸면 학습 때와 다른 기준선이 들어가 확률이 그럴듯하게 틀린다.
 */
function averageMlBaseline(samples: PostureFeatures[]): PostureMlFeatures | null {
  const ml = samples
    .map((f) => f.mlFeatures)
    .filter((m): m is PostureMlFeatures => m !== null);
  if (ml.length === 0) return null;

  const avg = (pick: (m: PostureMlFeatures) => number) =>
    ml.reduce((sum, m) => sum + pick(m), 0) / ml.length;

  return {
    eyeDistRatio: avg((m) => m.eyeDistRatio),
    earDistRatio: avg((m) => m.earDistRatio),
    noseAboveShoulder: avg((m) => m.noseAboveShoulder),
    earAboveShoulder: avg((m) => m.earAboveShoulder),
    noseZRel: avg((m) => m.noseZRel),
    earZRel: avg((m) => m.earZRel),
    acromionProxy: avg((m) => m.acromionProxy),
    headRoll: avg((m) => m.headRoll),
    shoulderTilt: avg((m) => m.shoulderTilt),
    absShoulderTilt: avg((m) => m.absShoulderTilt),
    absHeadRoll: avg((m) => m.absHeadRoll),
    shoulderZSpread: avg((m) => m.shoulderZSpread),
  };
}
